kore/src/Kore/Step/Simplification/AndTerms.hs

```diff
@@ -293,21 +309,45 @@ maybeTermAnd notSimplifier childTransformers first second = do
             lift $ Builtin.KEqual.unifyKequalsEq childTransformers notSimplifier unifyData
         | Just unifyData <- Builtin.KEqual.matchUnifyKequalsEq second first =
             lift $ Builtin.KEqual.unifyKequalsEq childTransformers notSimplifier unifyData
-        | otherwise =
-            asum
-                [ Builtin.KEqual.unifyIfThenElse childTransformers first second
-                , Builtin.Endianness.unifyEquals first second
-                , Builtin.Signedness.unifyEquals first second
-                , Builtin.Map.unifyEquals childTransformers first second
-                , Builtin.Set.unifyEquals childTransformers first second
-                , Builtin.List.unifyEquals
+        | Just ifThenElse <- Builtin.KEqual.matchIfThenElse first =
+            lift $ Builtin.KEqual.unifyIfThenElse childTransformers ifThenElse second
+        | Just ifThenElse <- Builtin.KEqual.matchIfThenElse second =
+            lift $ Builtin.KEqual.unifyIfThenElse childTransformers ifThenElse first
+        | Just unifyData <- Builtin.Endianness.matchUnifyEqualsEndianness first second =
+            lift $ Builtin.Endianness.unifyEquals first second unifyData
+        | Just unifyData <- Builtin.Signedness.matchUnifyEqualsSignedness first second =
+            lift $ Builtin.Signedness.unifyEquals first second unifyData
+        | Just unifyData <- Builtin.Map.matchUnifyEquals tools first second =
+            lift $ Builtin.Map.unifyEquals childTransformers tools first second unifyData
+        | Just unifyData <- Builtin.Map.matchUnifyEquals tools second first =
+            lift $ Builtin.Map.unifyEquals childTransformers tools second first unifyData
+        | Just unifyData <- Builtin.Set.matchUnifyEquals tools first second =
+            lift $ Builtin.Set.unifyEquals childTransformers tools first second unifyData
+        | Just unifyData <- Builtin.Set.matchUnifyEquals tools second first =
+            lift $ Builtin.Set.unifyEquals childTransformers tools second first unifyData
+        | Just unifyData <- Builtin.List.matchUnifyEqualsList tools first second =
```

Comment:
All of these `match...` functions should handle matching their arguments in both directions. They are the canonical matchers for each kind of unification problem and they should be symmetric.